### PR TITLE
Do not over-allocate when resizing in GeoHashTiler with bounds

### LINKED ISSUE #72521: [CI] GeoGridTilerTests testGeoHashBoundsExcludeTouchingTiles failing
Test can fail with an OOM exception.

**Build scan:**
https://gradle-enterprise.elastic.co/s/6bczyoza6ddji/tests/:x-pack:plugin:spatial:test/org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoGridTilerTests/testGeoHashBoundsExcludeTouchingTiles

**Reproduction line:**
`./gradlew ':x-pack:plugin:spatial:test' --tests "org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoGridTilerTests.testGeoHashBoundsExcludeTouchingTiles" -Dtests.seed=5211C01EDF07596D -Dtests.locale=uk -Dtests.timezone=America/Rio_Branco -Druntime.java=8`

**Applicable branches:**
master, 7.x

**Reproduces locally?:**
Yes

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoGridTilerTests&tests.test=testGeoHashBoundsExcludeTouchingTiles

**Failure excerpt:**
```
java.lang.OutOfMemoryError: Java heap space

  at __randomizedtesting.SeedInfo.seed([5211C01EDF07596D:ED5E1FFADCBA310]:0)
  at org.apache.lucene.util.ArrayUtil.growExact(ArrayUtil.java:323)
  at org.elasticsearch.index.fielddata.SortingNumericDocValues.growExact(SortingNumericDocValues.java:84)
  at org.elasticsearch.index.fielddata.SortingNumericDocValues.resize(SortingNumericDocValues.java:76)
  at org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoShapeCellValues.resizeCell(GeoShapeCellValues.java:45)
  at org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.AbstractGeoHashGridTiler.setValuesByRasterization(AbstractGeoHashGridTiler.java:108)
  at org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.AbstractGeoHashGridTiler.setValuesByRasterization(AbstractGeoHashGridTiler.java:101)
  at org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.AbstractGeoHashGridTiler.setValuesByRasterization(AbstractGeoHashGridTiler.java:101)
  at org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.AbstractGeoHashGridTiler.setValues(AbstractGeoHashGridTiler.java:45)
  at org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.BoundedGeoHashGridTiler.setValues(BoundedGeoHashGridTiler.java:17)
  at org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoShapeCellValues.advanceExact(GeoShapeCellValues.java:30)
  at org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoGridTilerTests.testGeoHashBoundsExcludeTouchingTiles(GeoGridTilerTests.java:248)
  at sun.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:498)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)

```

### PR DESCRIPTION
**Pull request by iverase**
Currently when declaring bounds with GeoHashTiler, we do not limit the amount of bucket we can create. This change prevents over allocating in such cases.

fixes #72521

### CODE REVIEW

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoGridTilerTests.java [hunk 1]
```diff
@@ -106,7 +106,7 @@ public void testGeoTileSetValuesBruteAndRecursivePoints() throws Exception {
     // tests that bounding boxes of shapes crossing the dateline are correctly wrapped
     public void testGeoTileSetValuesBoundingBoxes_BoundedGeoShapeCellValues() throws Exception {
         for (int i = 0; i < 1; i++) {
-            int precision = randomIntBetween(0, 4);
+            int precision = randomIntBetween(1, 4);
```
Comment (iverase) [inline x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoGridTilerTests.java:109]: I notice that this test might fail for precision 0. That precision is handle by AllCellValues so let's remove it.

Comment (imotov) [reply x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoGridTilerTests.java:109]: Do we have a separate test for that? Maybe it's better to pull GeoShapeCellIdSource in here and make sure that everything is still covered?

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHashGridTiler.java [hunk 2]
```diff
@@ -23,6 +23,9 @@
     /** check if the provided hash is in the solution space of this tiler */
     protected abstract boolean validHash(String hash);
 
+    /** Max size of the solution space */
+    protected abstract long getMaxHashes();
```
Comment (imotov) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHashGridTiler.java:27]: Would be nice to add a unit test for this one.

**Review by imotov — APPROVED** [review]
LGTM, left a couple of concerns regarding test coverage.

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoGridTilerTests.java (hunk 1 above)
Comment (iverase) [reply x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoGridTilerTests.java:109]: I am going to push it as it is. I handle this case in a separate PR as I think the current implementation is not right.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHashGridTiler.java (hunk 2 above)
Comment (iverase) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHashGridTiler.java:27]: done

**Pull request merged**